Bug fixes rules, see #413

ebaauw opened the pull request:
- Set `lasttriggered` in UTC instead of localtime;
- Fix firing rule with `in` condition on `/config/localtime`;
- Don't restore `timestriggered` from database on restart (so it resets to 0).

manup commented on rest_rules.cpp:
```diff
@@ -1350,35 +1350,23 @@ bool DeRestPluginPrivate::evaluateRule(Rule &rule, const Event &e, Resource *eRe
-            QTime rt;
 
-            if (rule.lastTriggered().isValid())
+            if (rule.lastTriggered().isValid() && rule.lastTriggered().secsTo(now) < 24 * 3600)
```
Does this work: `lastTriggered()` is UTC and `now` is local time?

ebaauw replied:
Funny enough it did.  Apparently the timezone is only considered when converting to/from a string.

Functionally it doesn't do what I wanted, thought, see below.

manup replied:
Interesting, maybe they also use the timezone info from QDateTime class and do some internal conversions.

ebaauw commented:
The fix for `in` condition on `/config.localtime` didn't work in full.  I changed the logic to no longer consider `lasttriggered`, but compare to the actual time (using `QTime::secsTo()` to discard the milliseconds).  This assumes `evaluateRule()` get called every second for `/config/localtime` - which it seems to be the case in my testing.  Might be safer to actually use the value for `/config/localtime` rather than `now` in these comparisons.

Also added support for `not in` and changed the `ddx` logic.  I saw some funny stuff that `ddx` rules fired incorrectly on creation, due to the `lasttriggered` logic.

ebaauw commented:
Damn, unintentionally included the change to simplify the code for the `lumi.sensor_86sw` switches in this PR as well.  Won't do any harm, though.

manup commented:
>The fix for in condition on /config.localtime didn't work in full. I changed the logic to no longer consider lasttriggered, but compare to the actual time (using QTime::secsTo() to discard the milliseconds). This assumes evaluateRule() get called every second for /config/localtime - which it seems to be the case in my testing. Might be safer to actually use the value for /config/localtime rather than now in these comparisons.

Yes I think that would be fine */config/localtime* is updated once every second and related indexed rules should be evaluated.

>Also added support for not in and changed the ddx logic. I saw some funny stuff that ddx rules fired incorrectly on creation, due to the lasttriggered logic.

:) cool thanks, I'll check it out soon, we use ddx heavily for the Phoscon App motion sensor controls. The not in might simplify some other rules too.

ebaauw commented:
> we use `ddx` heavily for the Phoscon App motion sensor controls. The `not in` might simplify some other rules too.

I use `ddx` extensively as well, in the logic to determine wether I left a room (motion in room, followed by motion in adjacent room, followed by no more motion in room).  Been running my latest commit in production since yesterday - it seems to work.

I've never used `not in` myself - it's functionally equivalent to `in` with the times reversed.  I did some basic tests, and it seems to work.

As mentioned elsewhere, to simplify rules, I would like an `in` (and `not in`) for numeric values as well (e.g. `state.status`, `state.buttonevent`).  Preferably they would take a range (like `"`_n_`/`_m_`"`) or a list (like `"`_n1_`,`_n2_`,`_n3_`"`).  My use case for a list would be a single rule to set `bri_inc` to 0 on long release of a Hue dimmer switch's dim up (2003) or dim down (3003).  
I also think we would benefit from `le` (<=) and `ge` (>=) operators, to make rules easier readable.  Of course, this is of less importance when the rules are generated by a script or by an app.
